### DISCUSSION TITLE
prevent 11-image not found from crashing eleventy

### DISCUSSION THREAD

**Discussion by btrem** [11ty/eleventy#2349] [Q&A] [answered]
I use 11ty-image to process/convert images with a shortcode to generate the markup. But if the image is missing -- like if I mistitle it or put it in the wrong place -- the image function fails, causing the npm 11ty script to fail. The code I have now:

    metadata = await image(rawImage, {
        widths: widths,
        formats: formats,
        outputDir: imageOutputDir,
        urlPath: imageUrlPath,
        filenameFormat: function (id, src, width, format, options) {
            const extension = path.extname(src)
            const name = path.basename(src, extension)
            return `${name}-${width}w.${format}`
          }
    });

I tried to use try/catch, but it still failed. I don't do enough programming to really understand try/catch, so I'm probably just getting it wrong. Can someone explain how to use it? Or some other way to prevent 11ty from crashing because an image file is missing?

**Comment by pdehaan** [comment, marked as answer]
@btrem Would something like this work?

```js
const Image = require("@11ty/eleventy-img");

async function imageShortcode(src, alt="", sizes, loading="lazy") {
  let metadata;
  try {
    metadata = await Image(src, {
      widths: [300, 600],
      formats: ["avif", "jpeg"]
    });
  } catch (err) {
    console.error(err.message);
    return "";
  }

  const imageAttributes = {
    alt,
    sizes,
    loading,
    decoding: "async",
  };

  // You bet we throw an error on missing alt in `imageAttributes` (alt="" works okay)
  return Image.generateHTML(metadata, imageAttributes);
}

module.exports = function(eleventyConfig) {
  eleventyConfig.addLiquidShortcode("image", imageShortcode);

  return {
    dir: {
      input: "src",
      output: "www",
    }
  };
};
```

```liquid
---
# src/index.liquid
title: 11ty-2349
---

<h1>{{ title }}</h1>

<div>
{% image "404.png", "missing logo" %}
</div>
```

### OUTPUT

```html
<h1>11ty-2349</h1>

<div>
</div>
```

And my Terminal output (macOS 12) looks like this:

```sh
> Executing task: npm run build <

> 11ty-2349@1.0.0 build
> eleventy

ENOENT: no such file or directory, stat '404.png'
[11ty] Writing www/index.html from ./src/index.liquid
[11ty] Wrote 1 file in 0.03 seconds (v1.0.1)
```

**Reply by pdehaan** [reply]
Although, reflecting back on this for a second, wrapping in a `try..catch` is a bit of a risky double-edged sword.

Now the site will seemingly build without throwing fatal errors, but if you don't see the log output, you might not notice there were some issues and some of the image paths were invalid; which might be fine for development, but you may want hard failures on a production build for a final site.

So I guess we could also add some more checks in the `imageShortcode()` function which checks if our `process.env.NODE_ENV` is "production" and if so, rethrow the error. Otherwise, assume it's a development build and quietly log the error to the console and move on.

**Comment by btrem** [comment]
Well, I'm a bit old school atm. I run 11ty on my macbook, then upload files to a shared hosting server. At some point, I need to move to netlify and then maybe have a build script connected to my github repo. It's not something I fully understand. But in the meantime, it's pretty safe to try...catch. I think. The only change is that I return an HTML comment in the catch block:

    catch (e) {
        console.log(e.message);
        return "";
    }
